### PR TITLE
[ID-2631] Expose zkEvm method for signing ejection transaction

### PR DESCRIPTION
**Pull request by matthewmuscat**
### Hi👋, please prefix this PR's title with:

- [ ] `breaking-change:` if you have introduced modification that necessitates immediate adjustments by this SDK's users to their applications, clients, or integrations to avert disruptions to existing features or functionalities.
- [ ] `feat:`, `fix:`, `refactor:`, `docs:`, or `chore:`.

# Summary
```ts
await provider.request({ method: 'im_signEjectionTransaction', params: [{
    from: "0x82163193c5d8d3fae054666125fce387ef5c16fb",
    to: "0x82163193c5d8d3fae054666125fce387ef5c16fb",
    chainId: "1",
    nonce: 123456789,
    value: '1',
}]})
```

# Detail and impact of the change

## Security

# Anything else worth calling out?

### CODE REVIEW

File: packages/passport/sdk/src/zkEvm/zkEvmProvider.ts [hunk 1]
```diff
@@ -464,6 +465,42 @@ export class ZkEvmProvider implements Provider {
...
+          );
+        }
+
+        const flow = trackFlow('passport', 'imSignEjectionTransaction');
+
+        try {
+          return await this.#guardianClient.withConfirmationScreen({
+            width: 480,
+            height: 720,
+          })(async () => {
```
Comment (haydenfowler) [inline packages/passport/sdk/src/zkEvm/zkEvmProvider.ts:483]: This flow shouldn't use the confirmation screen, so we can scrap this wrapping `withConfirmationScreen` call

File: packages/passport/sdk/src/zkEvm/transactionHelpers.ts [hunk 2]
```diff
@@ -212,3 +214,63 @@ export const prepareAndSignTransaction = async ({
...
+  zkEvmAddress,
+  flow,
+}: EjectionTransactionParams & { transactionRequest: TransactionRequest }) => {
+  const metaTransaction = await buildMetaTransactionForEjection(
+    transactionRequest,
+  );
+  flow.addEvent('endBuildMetaTransactions');
+
+  // NOTE: We sign again because we now are adding the fee transaction, so the
+  // whole payload is different and needs a new signature.
```
Comment (haydenfowler) [inline packages/passport/sdk/src/zkEvm/transactionHelpers.ts:265]:
```suggestion
```

**Comment by haydenfowler** [comment]
Can we add this new method to the sample app please? 
https://github.com/immutable/ts-immutable-sdk/blob/3e58d3c248504f819684c9987d6fd49d3581a044/packages/passport/sdk-sample-app/src/components/zkevm/Request.tsx#36

File: packages/passport/sdk/src/zkEvm/signEjectionTransaction.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  params: Array<any>;
+};
+
+export const signEjectionTransaction = async ({
+  params,
+  ethSigner,
+  zkEvmAddress,
+  flow,
+}: EthSendTransactionEjectionParams): Promise<EjectionTransactionResponse> => {
+  const transactionRequest = params[0] as TransactionRequest;
```
Comment (shirren) [inline packages/passport/sdk/src/zkEvm/signEjectionTransaction.ts:18]: @matthewmuscat not sure at which stage of the development cycle we are with these changes, but it'd be nice if we added a pre-condition check to this function call, so check `params` length perhaps.

File: packages/passport/sdk/src/zkEvm/transactionHelpers.ts [hunk 4]
```diff
@@ -212,3 +219,65 @@ export const prepareAndSignTransaction = async ({
...
+      'im_signEjectionTransaction requires a "chainId" field',
+    );
+  }
+
+  const metaTransaction: MetaTransaction = {
+    to: transactionRequest.to,
+    data: transactionRequest.data,
+    nonce: transactionRequest.nonce,
+    value: transactionRequest.value,
+    revertOnError: true,
```
Comment (shirren) [inline packages/passport/sdk/src/zkEvm/transactionHelpers.ts:252]: Any reason we do not set the `delegateCall`?

File: packages/passport/sdk/src/zkEvm/signEjectionTransaction.test.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+import { signEjectionTransaction } from './signEjectionTransaction';
+
+jest.mock('./transactionHelpers');
+jest.mock('../network/retry');
+
+describe('im_signEjectionTransaction', () => {
+  const signedTransaction = {
+    to: '123',
+    data: '123',
+    chainId: '123',
```
[the remark this answers is not in the document]
Comment (matthewmuscat) [reply packages/passport/sdk/src/zkEvm/signEjectionTransaction.test.ts:16]: for the test? or do you mean for the sample app?

File: packages/passport/sdk/src/zkEvm/transactionHelpers.ts (hunk 4 above)
Comment (matthewmuscat) [reply packages/passport/sdk/src/zkEvm/transactionHelpers.ts:252]: unsure myself, what benefit would we get from this? cc @haydenfowler if you have any opinion on this

File: packages/passport/sdk/src/zkEvm/signEjectionTransaction.test.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+import * as transactionHelpers from './transactionHelpers';
+import { signEjectionTransaction } from './signEjectionTransaction';
+
+jest.mock('./transactionHelpers');
+jest.mock('../network/retry');
+
+describe('im_signEjectionTransaction', () => {
+  const signedTransaction = 'signedTransaction123';
+
+  const transactionRequest: TransactionRequest = {
```
[the remark this answers is not in the document]
Comment (matthewmuscat) [reply packages/passport/sdk/src/zkEvm/signEjectionTransaction.test.ts:19]: yep fair call, thanks

File: packages/passport/sdk/src/zkEvm/transactionHelpers.ts (hunk 4 above)
Comment (haydenfowler) [reply packages/passport/sdk/src/zkEvm/transactionHelpers.ts:252]: @shirren I believe @matthewmuscat has followed the existing implementation in `eth_sendTransaction` here. As for _why_ `eth_sendTransaction` doesn't set `delegateCall`, I'm not too sure 😅 +1 to Matt's question above

**Review by haydenfowler — APPROVED** [review]

File: packages/passport/sdk/src/zkEvm/transactionHelpers.ts [hunk 7]
```diff
@@ -30,6 +30,13 @@ export type TransactionParams = {
   flow: Flow;
 };
 
+export type EjectionTransactionParams = Pick<TransactionParams, 'ethSigner' | 'zkEvmAddress' | 'flow'>;
+export type EjectionTransactionResponse = {
+  to: string;
+  data: string;
+  chainId: string;
+};
+
```
Comment (imx-mikhala) [inline packages/passport/sdk/src/zkEvm/transactionHelpers.ts:39]: If everything in this file is related to the ejection tx, any reason not to keep it inside `signEjectionTransaction.ts` ?

**Pull request merged**